blogpost/post.md

```diff
@@ -0,0 +1,261 @@
+original NumPy code. This is something, but it is not particularly impressive,
+given the speed-ups we have seen on CPU. Let's have a look into how to squeeze
+the most out of our GPU via a couple minor changes.
+
+**`float64` vs `float32`**. Many GPUs, in particular consumer-grade ones, are
+rather sluggish when running operations on `float64`. For this reason, changing
+the data generation to `float32`, the original NumPy code just gets a bit
+faster, about a 9%, but our CUDA code gets **40% faster**, yielding a **11x
+speed-up** over the plain NumPy code.
+
+`torch.compile`, by default, respects the NumPy semantics, and as such, it uses
+`np.float64` as its default dtype for all its creation ops. As discussed, this
+can hinder performance, so it is possible to change this default by setting
+
+```python
+from torch._dynamo import config
+config.numpy_default_float = "float32"
+```
+
+**CPU <> CUDA copies**. An 11x speed-up is good, but it is not even close to
+the CPU numbers. This is caused by a small transformation that `torch.compile`
+does behind the scenes. The code above takes NumPy arrays and returns NumPy
+arrays. All of these arrays are on CPU, but the computations are performed on
+the GPU. This means that every time the function is called, `torch.compile` has
+to copy all these arrays from CPU to the GPU, and then copy the result
+back to CPU to preserve the original semantics. There is no native
+solution to this issue in NumPy, as NumPy does not have the notion of a
+`device`. That being said, we can work around it by creating a wrapper to this
+function so that it accepts PyTorch tensors and returns PyTorch tensors.
+
+```python
+@torch.compile
+def tensor_fn(X, means):
+    X, means = X.numpy(), means.numpy()
+    ret = get_labels(X, means)
+    return torch.from_numpy(ret)
+
+def cuda_fn(X, means):
+    with torch.device("cuda"):
+        return tensor_fn(X, means)
+```
+
+This function now takes tensors in CUDA memory and returns tensors in CUDA
+memory, but the function itself is written in NumPy! When we keep the tensors
+in CUDA and perform the computations in `float32`, we see a **200x speed-up**
+over the initial NumPy implementation on `float32` arrays.
+
+**Mixing NumPy and PyTorch**. In this example, we had to write a small adaptor
+to convert tensors to ndarrays and then back to tensors. In programs that mix PyTorch and
+NumPy this is already done by calling `x.detach().cpu().numpy()` (or simply
+`x.numpy(force=True)`). Since when running under `torch.compile` we can run
+NumPy code in CUDA, we can simply modify this code to call `x.numpy()` and when
+running it under `device("cuda")`, as we did above, it will generate efficient
+CUDA code from original NumPy calls without copying the data from CUDA to CPU
+at all. Note that the resulting code would not run without `torch.compile`. For
+it to run in eager mode one would need to rollback to `x.numpy(force=True)`.
```

Comment:
This paragraph I guess explains the previous one, but still it's confusing, at least to not very torch-savvy reader in me. Consider rephrasing?